### DISCUSSION TITLE
About stateless and ContextListener, tokens in sessions

### DISCUSSION THREAD

**Discussion by pkly** [symfony/symfony#54230] [Q&A] [answered]
Hello.

I've recently upgraded our app from 5.4 to 6.4 and encountered an interesting problem.
While the configuration in 5.4 was kind of incorrect, it worked for our needs, as I'll explain in a second.

Our setup consists of the following:
- An authorization via JWT (Bearer token)
- Session data

Because of how symfony stores tokens in sessions via `Symfony\Component\Security\Http\Firewall\ContextListener` in 6.4 our flow broke, as previously logged in users via JWT and our custom authenticator would keep being logged on even after the token is no longer present in headers, as it was serialized via ContextListener.

This behavior was not present in 5.4 as we used the `stateless` property, but in 6.4 use of session while firewall is declared `stateless` causes an exception (which is fine, I don't mind this behavior). But then `ContextListener` will always serialize the token.

I managed to "fix" this by having a low priority listener on the response event and removing the `_security_firewall` key from session before it was saved, but I was wondering how would one go about this properly?

The only sensible resolution I can think of is to add another interface for storing security tokens and introduce that into `ContextListener`, then have it be configurable per-firewall and simply override the firewalls I don't need it at to a null one.

Any suggestions?

**Comment by MatTheCat** [comment, marked as answer]
Hello,

seems like you’re experiencing #48044. Fortunately since #49997 you can explicitly mark routes stateful to avoid this behavior.

You wouldn’t be the only one finding this is weird; there is still an open issue on this matter: #50715

**Reply by pkly** [reply]
I mean that sort of solves the issue?
So, should I mark firewalls as stateless: true, mark all my routes as stateless: false, and remove my `_security_` removal listener?

**Reply by MatTheCat** [reply]
Well for now it’s the “official” way to use the session in a request matching a stateless firewall, so I personally would do that.

Note that you can configure `stateless` for a [route group](https://symfony.com/doc/current/routing.html#route-groups-and-prefixes).

**Reply by pkly** [reply]
Yep yep, I knew that. I'm not going to add it by hand to 100+ routes, I'd go insane.

Thank you for the explanation and help.
